kpi-tracker/src/main.rs

```diff
@@ -0,0 +1,229 @@
+use std::collections::{BTreeMap, HashMap};
+use std::hash::Hash;
+
+use anyhow::Context;
+use chrono::{DateTime, Utc};
+use clap::Parser;
+use concordium_rust_sdk::{
+    id::types::AccountCredentialWithoutProofs,
+    smart_contracts::common::AccountAddress,
+    types::{
+        hashes::{BlockMarker, HashBytes},
+        AbsoluteBlockHeight,
+    },
+    v2::{self, AccountIdentifier, Client, Endpoint, FinalizedBlockInfo},
+};
+use futures::{self, StreamExt};
+
+#[derive(Debug, Parser)]
+struct Args {
+    /// The node used for querying
+    #[arg(long, default_value = "http://localhost:20001")]
+    node: Endpoint,
+    /// How many blocks to process.
+    // Only here for testing purposes...
+    #[arg(long, default_value_t = 10000)]
+    num_blocks: u64,
+}
+
+#[derive(Debug)]
+struct AccountDetails {
+    is_initial: bool,
+    block_hash: String, // FK to blocks
+}
+
+// Blocks are stored, so other tables can reference information about the block they were created in.
+#[derive(Debug)]
+struct BlockDetails {
+    block_time: DateTime<Utc>,
+    height: u64, // Used as a reference from where to restart on service restart.
+}
+
+type AccountsTable = HashMap<CanonicalAccountAddress, AccountDetails>;
+type BlocksTable = HashMap<String, BlockDetails>;
+
+struct DB {
+    blocks: BlocksTable,
+    accounts: AccountsTable,
+}
+
+#[derive(Eq, Debug, Clone, Copy, Ord, PartialOrd)]
+struct CanonicalAccountAddress(AccountAddress);
+
+impl From<CanonicalAccountAddress> for AccountAddress {
+    fn from(caa: CanonicalAccountAddress) -> Self {
+        caa.0
+    }
+}
+
+impl PartialEq for CanonicalAccountAddress {
+    fn eq(&self, other: &Self) -> bool {
+        let bytes_1: &[u8; 32] = self.0.as_ref();
+        let bytes_2: &[u8; 32] = other.0.as_ref();
+        bytes_1[0..29] == bytes_2[0..29]
+    }
+}
+
+impl Hash for CanonicalAccountAddress {
+    fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
+        let bytes: &[u8; 32] = self.0.as_ref();
+        bytes[0..29].hash(state);
+    }
+}
+
+impl AsRef<CanonicalAccountAddress> for AccountAddress {
+    fn as_ref(&self) -> &CanonicalAccountAddress {
+        unsafe { std::mem::transmute(self) }
+    }
+}
```

Comment:
For the love of god do not use `mem::transmute`. It is **wrong** in the current case  since you did not use `repr(transparent)`. It might work, but it is not guaranteed and new compiler versions can break the representation.

Reply:
Generally you should not need **any** unsafe code.